examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt

```diff
@@ -263,11 +268,51 @@ class MapboxNavigationActivity : AppCompatActivity() {
         }
     }
 
+    private val routesPreviewObserver = RoutesPreviewObserver { update ->
+        val routePreview = update.routesPreview
+        if (routePreview != null) {
+            routeLineAPI.setNavigationRoutes(
+                routePreview.routesList,
+                routePreview.alternativesMetadata
+            ) {
+                val style = mapboxMap.getStyle()
+                if (style != null) {
+                    routeLineView.renderRouteDrawData(style, it)
+                }
+            }
+            // update the camera position to account for the new route
+            viewportDataSource.onRouteChanged(routePreview.primaryRoute)
+            viewportDataSource.evaluate()
+
+            binding.mapView.gestures.addOnMapClickListener(previewMapClickListener)
```

Comment:
another option is to put a guard in the previewMapClickListener

```
  val routePreview = MapboxNavigationApp.current()?.getRoutePreview() ?: return
```

Reply:
preference typically to attach listeners to a lifecycle and then handle the logic in the click listener. rather than creating logic that controls the click listener.

reason is because it's easy to leak subscriptions if logic is controlling them. for example, the route preview emit may not happen again. so you will need to removeOnMapClickListener when the lifeycycle is destoyed.

it doesn't apply to this activity because the mapView and the subscriptions are destroyed with the activity

Reply:
nice idea! now listener is always registered but I check route preview state inside.